.github/workflows/docker.yml

```diff
@@ -15,11 +15,17 @@ on:
     branches:
       - master
     paths: ['.github/workflows/docker.yml', '.devops/*.Dockerfile', '**/CMakeLists.txt', '**/Makefile', '**/*.h', '**/*.hpp', '**/*.c', '**/*.cpp', '**/*.cu', '**/*.cuh', '**/*.swift', '**/*.m', '**/*.metal']
+  workflow_dispatch: # allows manual triggering, useful for debugging
 
 concurrency:
   group: ${{ github.workflow }}-${{ github.head_ref && github.ref || github.run_id }}
   cancel-in-progress: true
 
+# Fine-grant permission
+# https://docs.github.com/en/actions/security-for-github-actions/security-guides/automatic-token-authentication#modifying-the-permissions-for-the-github_token
+permissions:
+  packages: write
```

Comment:
@ggerganov Small explanation for this: This allow disabling global write access for `GITHUB_TOKEN` on CI.

However, **some other workflows** don't support this yet (for example, `build.yml`). I'll do a follow-up PR for it.

![image](https://github.com/user-attachments/assets/b4f57264-68c3-4400-8566-491534c08127)